doc/compiler/optimizer/InliningAndInliningGuards.md

```diff
@@ -0,0 +1,406 @@
+
+## Nopable Guard
+
+Nopable guard is like a conditional test conceptually except there is no explicit testing in the generated code.
+There is an `“if”` in IL, but it will always fall into the inlined code during the compilation. If the user
+redefines the method that is inlined in the future, we will go patch the code where it is inlined and take the
+slow path to invoke the new version of the method. Such events are stop-the-world events where application
+threads are guaranteed to not be executing, and hence patching in a full instruction into the instruction
+stream is safe.
+
+For a nopable guard to be used, it depends on the state of the class hierarchy at compile time. For example, `B`
+subclasses an interface `A` when the compilation is done. `A` declares a method `foo()` which is implemented by
+`B`. When the method is compiled and inlined, there is only one target `B`. In the future, if a subclass `C` is
+loaded after the compilation is done and the code has been generated, there could be an object type of `B` as
+well as type `C` that goes through `A.foo()`. Once `C` gets loaded, nopable guards cannot be used any more.
+Before `C` gets loaded, JIT has an infrastructure that receives a notification that a new class is being loaded.
+It assesses if any of the guards will be invalidated. We will patch the code where we have the nopable guards to
+go to the slow path to invoke the call. Afterwards, the class `C` will be allowed to be loaded and to enter the
+system. When we do nopable guard patching, the guard changes the state. We convert an unconditional fall through
+into an unconditional goto.  Even if it is the object `B` that flows through, it will still jump to the slow
+path to invoke the virtual call. Nopable guards require JIT to monitor the changes of program behaviour such as
+class loading/unloading and do the patching to ensure inlined code is not invoked when we are not allowed to.
+
+## Explicit Guard
+
+Explicit guards are the guards that we generate the code for what we have in the trees.
+
+There are two different cases. One case is that there is [a bit in every J9Method](https://github.com/eclipse-openj9/openj9/blob/8393a7277cbd664acc8c9477d66b2dc9a9d6f412/runtime/compiler/env/j9method.cpp#L5074-L5077)
+indicating whether the method is overridden or not.  Any time a subclass is loaded, the VM checks if the method
+in that class overrides the method in the super class and it sets that bit for that method. We can always
+inline the method with a guard that checks the method. We explicitly test that bit in our IL trees. If it is not
+overridden, the inline code is executed, otherwise it invokes the virtual call.
+
+Another case of explicit test is based on profiling which is more common by default. Profiler tracks what type
+of objects we encounter when we have the virtual call or the interface call. That is the receiver type.
```

Comment:
"That is the receiver type." seems like it should be merged with some other sentence or deleted